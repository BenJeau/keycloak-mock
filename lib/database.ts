import { v4 as uuidv4 } from "uuid";

export interface CreateMockUserProfileOptions {
  name: string;
  email: string;
}

export interface MockUserProfile {
  sub: string;
  name: string;
  email: string;
  gender: "male" | "female";
  preferred_username: string;
  given_name: string;
  family_name: string;
  email_verified: boolean;
}

class MockDatabase {
  users: MockUserProfile[];

  constructor() {
    this.users = [];
  }

  findUserByID(sub: string): MockUserProfile | null {
    return this.users.find((storedUser) => storedUser.sub === sub) || null;
  }

  clear(): void {
    this.users = [];
  }

  /**
   * Creates a new user and returns the profile of the newly created user.
   */
  createUser(options?: CreateMockUserProfileOptions): MockUserProfile {
    const name = options ? options.name : "Henk Jansen";
    const email = options ? options.email : "henk@gmail.com";

    const [givenName, familyName] = name.split(" ");

    const profile: MockUserProfile = {
      sub: uuidv4(),
      email,
      name,
      email_verified: true,
      gender: "male",
      given_name: givenName,
      family_name: familyName,
      preferred_username: email,
    };

    this.users.push(profile);

    return profile;
  }
}

export default MockDatabase;
