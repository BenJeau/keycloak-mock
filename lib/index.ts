export {
  Mock,
  MockOptions,
  activateMock,
  deactivateMock,
  getMock,
  getMockInstance,
} from "./mock";
export {
  MockInstance,
  MockInstanceParams,
  CreateMockInstanceOptions,
  createMockInstance,
} from "./instance";
export {
  default as MockDatabase,
  MockUser,
  MockUserCredential,
  MockUserCredentialType,
  MockUserProfile,
  CreateMockUserOptions,
} from "./database";
export {
  default as createBearerToken,
  CreateTokenOptions,
} from "./createBearerToken";
