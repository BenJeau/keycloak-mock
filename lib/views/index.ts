export { default as listCertificates } from "./listCertificates";
export { default as getUserInfo } from "./getUserInfo";
export { ViewFn } from "./types";
