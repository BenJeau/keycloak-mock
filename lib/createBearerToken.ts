import { v4 as uuidv4 } from "uuid";
import jwt from "jsonwebtoken";

import { JWK } from "node-jose";

import { MockUserProfile } from "./database";

export interface CreateTokenOptions {
  user: MockUserProfile;
  expiresIn: number;
  key: JWK.Key;
  realm: string;
  clientID: string;
  authServerURL: string;
}

const createBearerToken = (options: CreateTokenOptions): string => {
  const timestamp = Math.floor(Date.now() / 1000);
  const expiresAt = timestamp + options.expiresIn;

  return jwt.sign(
    {
      iss: `${options.authServerURL}/realms/${options.realm}`,
      iat: expiresAt,
      exp: expiresAt,
      nbf: 0,
      typ: "Bearer",
      sub: options.user.sub,
      azp: options.clientID,
      session_state: uuidv4(),
      gender: options.user.gender,
      preferred_username: options.user.preferred_username,
      given_name: options.user.given_name,
      family_name: options.user.family_name,
      email_verified: options.user.email_verified,
      name: options.user.name,
      email: options.user.email,
    },
    options.key.toPEM(true),
    {
      algorithm: "RS256",
      header: {
        typ: "JWT",
        kid: options.key.kid,
      },
    }
  );
};

export default createBearerToken;
