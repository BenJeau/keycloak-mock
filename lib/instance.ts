import { JWK } from "node-jose";

import createBearerToken from "./createBearerToken";
import MockDatabase, { MockUserProfile, CreateMockUserProfileOptions } from "./database";

export interface CreateMockInstanceOptions {
  authServerURL: string;
  clientID: string;
  realm: string;
}

export interface MockInstanceParams {
  authServerURL: string;
  clientID: string;
  realm: string;
}

class MockInstance {
  store: JWK.KeyStore;
  defaultKey: JWK.Key;
  database: MockDatabase;
  params: MockInstanceParams;

  constructor(
    store: JWK.KeyStore,
    defaultKey: JWK.Key,
    database: MockDatabase,
    params: MockInstanceParams
  ) {
    this.store = store;
    this.defaultKey = defaultKey;
    this.database = database;
    this.params = params;
  }

  createURL(path: string): string {
    return `${this.params.authServerURL}${path}`;
  }

  createBearerToken(sub: string, expiresIn: number = 3600): string {
    const user = this.database.findUserByID(sub);
    if (!user) {
      throw new Error("Cannot create bearer token for non-existen user");
    }

    return createBearerToken({
      user,
      key: this.defaultKey,
      expiresIn,
      realm: this.params.realm,
      clientID: this.params.clientID,
      authServerURL: this.params.authServerURL,
    });
  }
}

const createMockInstance = async (options: CreateMockInstanceOptions): Promise<MockInstance> => {
  const store = JWK.createKeyStore();
  const defaultKey = await store.generate("RSA", 2048, { use: "sig" });

  return new MockInstance(store, defaultKey, new MockDatabase(), {
    authServerURL: options.authServerURL,
    clientID: options.clientID,
    realm: options.realm,
  });
};

export { MockInstance, createMockInstance };
