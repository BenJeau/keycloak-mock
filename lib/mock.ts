import nock, { Scope } from "nock";

import { MockInstance } from "./instance";

import { ViewFn, listCertificates, getUserInfo } from "./views";

export interface Mock {
  scope: Scope;
}

export interface MockOptions {
  listCertificatesView?: ViewFn;
  getUserInfoView?: ViewFn;
}

const activateMock = (instance: MockInstance, options?: MockOptions): Mock => {
  const scope = nock(instance.params.authServerURL)
    .persist()
    .get(`/realms/${instance.params.realm}/protocol/openid-connect/certs`)
    .reply(function() {
      if (options && options.listCertificatesView) {
        return options.listCertificatesView(instance, this.req);
      }

      return listCertificates(instance, this.req);
    })
    .get(`/realms/${instance.params.realm}/protocol/openid-connect/userinfo`)
    .reply(function() {
      if (options && options.getUserInfoView) {
        return options.getUserInfoView(instance, this.req);
      }

      return getUserInfo(instance, this.req);
    });

  return { scope };
};

const deactivateMock = (mock: Mock): void => {
  mock.scope.persist(false);

  // @ts-ignore
  mock.scope.keyedInterceptors = {};

  // @ts-ignore
  mock.scope.interceptors = [];
};

export { activateMock, deactivateMock };
