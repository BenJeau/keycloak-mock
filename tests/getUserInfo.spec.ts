import axios from "axios";

import * as KeycloakMock from "../lib";
import { setupBefore, teardownAfter, getMockInstance } from "./util";

describe("getUserInfo", () => {
  beforeAll(setupBefore);
  afterAll(teardownAfter);

  it("works", async () => {
    const kmock = getMockInstance();

    const user = kmock.database.users[0];
    const token = kmock.createBearerToken(user.sub);

    const url = kmock.createURL("/realms/myrealm/protocol/openid-connect/userinfo");

    const response = await axios.get(url, { headers: { authorization: `Bearer ${token}` } });

    // hack out created_at because it keeps changing
    const responseData = { ...response.data, created_at: 1 };
    expect(responseData).toMatchSnapshot();
  });
});
