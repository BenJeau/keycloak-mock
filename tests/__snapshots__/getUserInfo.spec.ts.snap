// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getUserInfo works with token 1`] = `
Object {
  "created_at": 1,
  "email": "henk@gmail.com",
  "email_verified": true,
  "family_name": "Jansen",
  "gender": null,
  "given_name": "Henk",
  "name": "Henk Jansen",
  "preferred_username": "henk.jansen@gmail.com",
  "sub": "4255c2ae-3f11-4a26-a401-3981c8845df3",
}
`;
